### ISSUE TITLE
Discussion: Finalize API and Ember 2.0 support

### ISSUE THREAD

**Issue by jacojoubert** [emberui/emberui#103]
It is time to update EmberUI to support Ember 2.0 and bring EmberUI itself to version 1.0. A lot has changed since EmberUI was first started, and it is time to advantage of some of the new conventions to make it more customizable and flexible.

**The purpose of this post is to gather feedback on the proposed changes.**

#### Goals

Ember 2.0 compatibility
Finalized API for EmberUI that is flexible and extensible
Move to ES6 from Coffeescript

## Poplist (placeholder name)

This would promote the `poplist` component for for public use. Making use of the new `ember-collections` component to replace `list-view`, we will allow for the customization of each of the items in the poplist. This would allow the popup to show a great variety of content, not just a simple text list.

    {{#eui-poplist
      options=options
      target="button-id"
      targetAttachment="top left"
      onClose="poplistClosed" as |option index|}}
        {{option.label}}
    {{/eui-poplist}}

The popup can be positioned relative to an element via the `target` attribute. You pass it an `id` selector of the element you want to position against, as well as the `attachment` and `targetAttachment`. We will use Tether to do the actual positioning.

## Button

The `button` will be converted to a block component. This would drop the `label` attribute and instead allow the user to customize the content of the button itself however they choose. This would make the usage identical to the `<button>` element from HTML.

    {{#eui-button
      action="save"
      style="primary"
      size="medium"}}
        Label of button
    {{/eui-button}}

The `leadingIcon` and `trailingIcon` attributes would be removed, and instead it would be up to the user to insert a tag inside the block that displays the icon. This would allow for css based icons, or using SVGs and other images.

## Dropbutton

The `dropbutton` would be converted to a block element and take the styling options from the `button`. You can customize how each of the options are displayed via the contents of the block.

    {{#eui-dropbutton
      label="Actions"
      options=options as |option index|}}
        {{eui-dropbutton-option label=option.label click=option.action}}
    {{/#ui-dropbutton}}

A default `dropbutton-option` component will be provided that provides the same functionality as the current `dropbutton`. The `dropbutton` would also retain the `label` attribute to customize the label of the button when there is no primary action.

## Select

The `select` will be converted to a block element and take the styling options from the `button`. You can customize how each of the options are displayed via the contents of the block.

    {{#eui-select
      placeholder="Select a country"
      selection=selection
      value=value
      valuePath="value"
      onChange="save"
      options=options as |option index|}}
        {{eui-select-option label=option.label}}
    {{/eui-select}}

A default `eui-select-option` component will be provided that provides the same functionality as the current select.

You can specify an action via `onChange` that needs to fire when the selection changes.

## Checkbox

The `checkbox` will be converted to a block element to allow for the display of the label and consistency with the other components.

    {{#eui-checkbox
      value=rememberMe}}
        Remember me
    {{/eui-checkbox}}

## Input & Textarea

The `input` and `textarea` components will become wrappers around the ember components. The `placeholder` attribute would also be removed and done via the native placeholder attribute. This means IE8 will no longer be supported.

    {{#eui-input
      value=value
      error=error
      forceErrorCheck=true}}
        {{input type="text placeholder="Email address" value=value}}
    {{/eui-input}}

By separating the input out it will be easier to write custom inputs (for example that only allows certain kinds of inputs) and layer the EmberUI error handling and styling on top.

    {{#eui-input
      value=value
      error=error}}
        {{custom-input type="text" value=value}}
    {{/eui-input}}

## Things that still needs attention

- Animations aren't done very well right now. Changes should be made, just not sure how to structure it yet.
- Specifying the label via an attribute on `dropbutton` is inconsistent with other components. Not sure if there is another approach we could take. 
- Does the `calendar` and `select-date` components need any changes?

**Comment by jacojoubert** [comment]
Pinging @ghedamat @elucid @artsyca @kenips 

Would love some feedback on the proposed changes.

**Comment by ghedamat** [comment]
@jacojoubert added to the review list :)
will try to get to this shortly

**Comment by artsyca** [comment]
Me to @jacojoubert thanks for including me!

**Comment by kenips** [comment]
:+1: love it overall! I agree that `eui-dropbutton` is the most weird of all especially requiring `eui-dropbutton-option` to function. Removing use of label everywhere is nice.

Another suggestion is to start removing two-way bindings - while fixing compatibility for 1.13 I had already removed reliance on two-way binding for `selection` between `eui-calendar` and `eui-popcal`. (see d947058). The use of `Ember.run.next` [in current master](https://github.com/emberui/emberui/blob/master/addon/components/eui-calendar.coffee#L72) is a smell to me. The use of two-way binding is particular problematic with calendar's `selection` which deals with moment object and array of moment of objects depending on the state.

**Comment by jacojoubert** [comment]
So it doesn't strictly require the `dropbutton-option` component, but it does make it easier. If you want to customize how the popup items look for your `dropbutton` though you will replace it with your own custom component. The same applies for the `select` component.

I agree that two-way bindings should be removed, unless it is a form component such as an input or select that has to send their data back to the app.

**Comment by taras** [comment]
@jacojoubert Overall, I agree with the direction. 

We could consider making DDAU single API because it's easy to opt in to TwoWayBinding like interface by using `{{mut}}` helper.

**Comment by jacojoubert** [comment]
Do you have an example of what you mean?

**Comment by taras** [comment]
Sorry, it took so long to get back to you on this, but it's basically this

```
<input type="text" value={{text}} oninput={{action (mut text) value="target.value"}}
```

**Comment by ghedamat** [comment]
@taras I agree, I think this will look/work amazing, but until kebab-actions are implemented I'm afraid we should not got for this approach given that it has potential performance problems that might be even more evident on a component library like this one.

reference:
https://github.com/emberjs/rfcs/pull/100

**Comment by taras** [comment]
For now, our biggest concern is the API that included addons will provide, for example

```
{{event-calendar selection={{selectedEvent}} select="selectEvent"}}
```

will do the same thing as 

```
{{event-calendar selection={{selectedEvent}} select=(action (mut selectedEvent))}}
```

We can do ^^ today without anything related to kebab actions.

**Comment by ghedamat** [comment]
yup! for this one I fully agree, I was just concerned about native events :)

**Comment by kenips** [comment]
Also now that we're in ES6, might be good to look at Dockyard's [styleguide](https://github.com/dockyard/styleguides/blob/master/engineering/ember.md). In particular [creating local versions of Ember modules](https://github.com/dockyard/styleguides/blob/master/engineering/ember.md#create-local-version-of-ember-and-ds) is worth adopting imo.

One additional request: getting rid of global moment. [ember-moment](https://github.com/stefanpenner/ember-moment) has some nice CP as well as ES6 moment for importing (via [ember-cli-moment-shim](https://github.com/jasonmit/ember-cli-moment-shim)).

**Comment by kenips** [comment]
@jacojoubert base on the current WIP for `eui-button`, are we dropping loading state support?

**Comment by jacojoubert** [comment]
@kenips No, loading state is there and it works. I will see if I can find time this week to wrap things up.

**Comment by kenips** [comment]
:+1: :heart: i'm good to test this weekend.

**Comment by jacojoubert** [comment]
Sorry about the delay in this. Will try to get it done by next weekend. There is quite a bit of work left on updating the docs and website.

Also a decision has to be made on whether ember-collection or smoke-and-mirrors is the best library to use for the options component.

**Comment by kenips** [comment]
That's cool - thanks for the update. I'll get started integrating this weekend and see if I run into any issues.

Haven't been following the two but in general I'd go with "official". Any problem with ember-collection?

**Comment by jacojoubert** [comment]
Two reasons

1. ember-collection has been incredibly slow to release. It still doesn't have a published addon according to their readme.
2. smoke-and-mirrors support flexible row heights. So each item in a select doesn't have to be the same height, and the user doesn't need to specify the height up front. This will make for a much easier customizable select.

**Comment by kenips** [comment]
I took this to Ember Slack and core team didn't comment on the validity of ember-collection, but:

![collection-or-smoke](https://cloud.githubusercontent.com/assets/794915/11448251/73090152-9522-11e5-80d1-18f9d27ed4c3.png)

**Comment by jacojoubert** [comment]
So I tried out smoke-and-mirrors, and it still renders a single div for each option. They are empty, but doing that for a large list makes it unusable on super large lists. So it isn't an option if that doesn't get fixed.

I would prefer to use none of them, so I am thinking maybe the list needs to be optional. I can provide a very simple list, and then allow the user to swap out the list with their own. The API to override the default list would be like so:

```
{{#eui-select 
  listComponent=(component "custom-list-component)
  options as |option|}
    {{option.label}}
{{/eui-select}}
```
